### ISSUE TITLE
[request] speed-up multidim slicing backward

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#9701]
The current way of apply slicing (indexing) isn't very efficient ([here](https://github.com/pytorch/pytorch/blob/7160846c81d9ebd298a4c1608d67b8f4bd76fee7/torch/csrc/autograd/python_variable_indexing.cpp#L151-L199)). Each slicing at each dim is applied as a separate autograd op, making the backward extremely inefficient:
```py
>>> x = torch.randn(3,3,requires_grad=True)
>>> y = x[0, 1]
>>> y.grad_fn
<SelectBackward object at 0x7fbca03e0668>
>>> y.grad_fn.next_functions  # two <SelectBackward>s
((<SelectBackward object at 0x7fbca03e0550>, 0),)
>>> y.grad_fn.next_functions[0][0].next_functions
((<AccumulateGrad object at 0x7fbca03e0668>, 0),)
```
The backward of this would be:
```
   gy      = [1]
=> g(x[0]) = [0, 1, 0]
=> g(x)    = [[0, 1, 0], [0, 0, 0], [0, 0, 0]]
```
, incurring extra copying and malloc'ing. This will be especially slow with large tensors.

Potential solutions: 
1. Use `as_strided` in `applySlicing`.
  pros: fits the current structure easily.
  cons: the backward isn't well optimized for slicing (but it's not terrible either if input is not expanded...).
2. Move multidim slicing to ATen.
  pros: benefit cpp API
  cons: need ways to represent `3..5` and `...`.

cc @colesbury

**Comment by srush** [comment]
Any progress or suggestions for getting around this issue? 

My backward pass is being dominated by a bunch of SliceBackwards.

SliceBackward                                13.90%        120.106ms           34.89%        301.530ms        108.347us           24.19%        327.914ms        117.827us             2783
zeros                                        11.02%         95.243ms           24.42%        211.019ms         53.020us           14.37%        194.825ms         48.951us             3980
zero_                                        10.18%         87.965ms           10.18%         87.965ms         20.106us           11.03%        149.487ms         34.169us             4375
add                                           3.54%         30.571ms            3.54%         30.571ms         27.995us           10.89%        147.641ms        135.203us             1092
SelectBackward                                7.34%         63.480ms           19.22%        166.155ms        138.810us           10.47%        141.884ms        118.533us             1197
